### PR TITLE
improve: apply `tachyon` optimizations(2 & 5)

### PR DESCRIPTION
**Pull request by guorong009**
## Description
Analyze and apply the `tachyon` optimizations - **2(batch normalize)** & **5(save heap allocations)**

## Related issues
  - #293 

## Changes
  - apply the `batch_normalize` in everywhere possible of `halo2_backend` crate 
 
## NOTE
1. `batch_normalize`
In order to apply the `batch_normalize` optimization, the algorithm of `batch_normalize` & `batch_inverse` should be implemented for elliptic curves. It is already implemented in [pse/halo2curves](https://github.com/privacy-scaling-explorations/halo2curves/blob/main/src/derive/curve.rs#L775-L806) repo.

2. `save heap allocations`
For this optimization, I just apply one - remove unnecessary "clone" op in ["permutation_commit"](halo2_backend/src/plonk/permutation/prover.rs).
The reason is that other optimizations needless in Rust project. 
Also, the `Polynomial` implementation has big diff between projects.

### CODE REVIEW

**Review by ed255 — APPROVED** [review]
The changes LGTM!
I think the most significant one may be the memory optimization.
For the batch normalization, I believe the total cost saving is very small when done after an MSM because the MSM is already very costly (the MSM overshadows the improvement)
